### DISCUSSION TITLE
how to improve the performance of group by  with tens of billions of data use trino

### DISCUSSION THREAD

**Discussion by yx-keith** [trinodb/trino#18168] [Q&A]

**Comment by jiangzhx** [comment]
This depends on many possibilities.
for example: 

1. group by on high cardinality or low cardinality.
2. group by key already bucketed or not.

**Reply by yx-keith** [reply]
sql like that:

**select count(*) from (
select name,count(*) from hive.test.test_orc where dt >='20230206' and dt<='20230207' group by name
) t;**

result of this query is tens of millions,  I think this is on high cardinality
then key is not bucketed

**Reply by yx-keith** [reply]
trino cluster is 448 cpu cores and 2.1TB memory, 7 worker nodes.

Elapsed time is 50s-1min
input data is 164GB

can be faster?

**Reply by jiangzhx** [reply]
1.Try to bucket group by key "name". The bucket number should be greater than the number of worker nodes (7). 
Doing this will provide benefits from the "bucket_execution_enabled" feature, which will reduce the cost of "Internal Network Data" .

2.Sort data with the column "name". 
Regarding this, I am not 100% sure. Does Trino have the ability of "streaming aggregation"? It will reduce memory usage.

3.does column "dt" was partition column? 

4.how about file size in hdfs?

**Reply by yx-keith** [reply]
does bucket group by key "name" means  to bucket table when create hive table?

yes, dt is partition column.

**Reply by jiangzhx** [reply]
> does bucket group by key "name" means  to bucket table when create hive table?
> 
> 
> 
> yes, dt is partition column.

yes

**Reply by yx-keith** [reply]
file size is 3.4TB , number of files is 3728

**Reply by yx-keith** [reply]
most of time(80%) is spent in source stage,  wall time is 1.73 times of cpu time in source stage, does it mean that  most of time is spent on I/O?
